### PR TITLE
[terra-core] Allow hover styles to apply only on non-touch devices.

### LINKED ISSUE #2927: Investigate whether Hover highlight is enabled on clicking using mobile devices
# Bug Report

## Description
Investigate all the terra-core components where Hover styles are implemented. such that in mobile devices when performing actions other than hovering on that particular component  whether the hover color is highlighted.

Related to #2891 and #2076 

## Expected Behavior
Hover styles should be only applied during hover action

## Possible Solution
Similar to this [PR](https://github.com/cerner/terra-core/pull/2332)

## @ Mentions
@neilpfeiffer

### PR DESCRIPTION
**Pull request by saket2403**
### Summary
This PR resolves the issue by making use of hover and pointer interaction media features introduced in Media Queries Level 4 so that the hover styles are applied only on non-touch devices whose primary input is a fine pointer device such as a mouse or a stylus.

Closes #2927 

### Deployment Link

https://terra-core-deployed-pr-#.herokuapp.com/

### Additional Details

Based on fix introduced for terra-list: #2332 
and terra-button: #2929
Reverts the wdio changes in #2929 after applying ontouchstart instead of media query.

### CODE REVIEW

**Comment by ShettyAkarsh** [comment]
@saket2403  can you share the before and after screenshots here ? for all the components where the changes were made

**Comment by tbiethman** [comment]
@neilpfeiffer @jeremyfuksa Before I sign off on this, can you guys confirm this is actually something we want to do? I know this originated from a number of logged issues, but I don't see anything that traces back directly to one of you guys.

Based on previous usage elsewhere, I expect all this to function okay. But our reliance on non-standard media features here, and our general hesitance to code against standard browser behaviors, gives me pause.

**Comment by supreethmr** [comment]
are we tested this changes in IE....?

**Comment by saket2403** [comment]
Changes reverted to use media queries and wdio tests removed for hover styles.

**Comment by supreethmr** [comment]
we might need to add these changes for terra-switch.

**Comment by saket2403** [comment]
> we might need to add these changes for terra-switch.

Done 1bd14b6

File: packages/terra-paginator/src/Paginator.module.scss [hunk 1]
```diff
@@ -46,12 +46,15 @@
       white-space: normal;
       word-wrap: break-word; /* For IE 10 and IE 11 */
 
-      &:hover {
-        background-color: var(--terra-paginator-nav-link-hover-background-color, transparent);
-        border: var(--terra-paginator-nav-link-hover-border, none);
-        box-shadow: var(--terra-paginator-nav-link-hover-box-shadow);
-        color: var(--terra-paginator-nav-link-hover-color, #001f67);
-        text-decoration: var(--terra-paginator-nav-link-hover-text-decoration, underline);
+      /* stylelint-disable max-nesting-depth */
```
Comment (emilyrohrbough) [inline packages/terra-paginator/src/Paginator.module.scss:49]: Can we disable next line or re-enable this rule after this query?

File: packages/terra-show-hide/src/_ShowHideButton.module.scss [hunk 2]
```diff
@@ -66,14 +66,17 @@
       }
     }
 
-    &:hover,
-    &.is-hovered {
-      background-color: var(--terra-show-hide-v1-button-hover-background-color);
-      color: var(--terra-show-hide-v1-button-hover-color, #001f67);
-      text-decoration: var(--terra-show-hide-v1-button-hover-text-decoration, underline);
+    /* stylelint-disable max-nesting-depth */
```
Comment (emilyrohrbough) [inline packages/terra-show-hide/src/_ShowHideButton.module.scss:69]: same question as above

File: packages/terra-table/src/subcomponents/HeaderCell.module.scss [hunk 3]
```diff
@@ -45,12 +45,16 @@
...
 
+    @media screen and (-ms-high-contrast: active), (-ms-high-contrast: none), (hover: hover) and (pointer: fine) {
+      &:hover {
+        background-color: var(--terra-table-header-cell-hover-background-color, #f1f1f2);
+        background-image: var(--terra-table-header-cell-hover-background-image, linear-gradient(to bottom, #d1d2d3 2%, #e0e1e2 30%, #ebebeb 80%));
+        box-shadow: var(--terra-table-header-cell-hover-box-shadow, inset 0 0 0 1px #f1f1f2);
+      }
+    }
+      
+    &:hover {
```
Comment (emilyrohrbough) [inline packages/terra-table/src/subcomponents/HeaderCell.module.scss:57]: Is this right? Did this need to be wrapped in a media query?

File: packages/terra-table/src/subcomponents/Row.module.scss [hunk 4]
```diff
@@ -102,18 +106,20 @@
...
-      background-image: var(--terra-table-row-selected-hover-background-image, linear-gradient(to bottom, #b3ddf8 2%, #c8e9fe 30%, #d1ebfb 80%));
-      box-shadow: var(--terra-table-row-selected-hover-box-shadow, inset 0 0 0 1px #e1f3fd);
+      &.is-selected:hover {
+        background-color: var(--terra-table-row-selected-hover-background-color, #d1ebfb);
+        background-image: var(--terra-table-row-selected-hover-background-image, linear-gradient(to bottom, #b3ddf8 2%, #c8e9fe 30%, #d1ebfb 80%));
+        box-shadow: var(--terra-table-row-selected-hover-box-shadow, inset 0 0 0 1px #e1f3fd);
+      }
     }
-
+
```
Comment (emilyrohrbough) [inline packages/terra-table/src/subcomponents/Row.module.scss:122]: extra spacing

File: packages/terra-paginator/src/Paginator.module.scss (hunk 1 above)
Comment (saket2403) [reply packages/terra-paginator/src/Paginator.module.scss]: Done 9892dee

File: packages/terra-show-hide/src/_ShowHideButton.module.scss (hunk 2 above)
Comment (saket2403) [reply packages/terra-show-hide/src/_ShowHideButton.module.scss]: Done 9892dee

File: packages/terra-table/src/subcomponents/HeaderCell.module.scss (hunk 3 above)
Comment (saket2403) [reply packages/terra-table/src/subcomponents/HeaderCell.module.scss]: Ahh missed it, added here 9892dee

File: packages/terra-table/src/subcomponents/Row.module.scss (hunk 4 above)
Comment (saket2403) [reply packages/terra-table/src/subcomponents/Row.module.scss]: Removed 9892dee

**Comment by neilpfeiffer** [comment]
Tested: PR #2932
Resolves: Issue #2927

**Reviewed:**
Manually reviewed the hover states for the components affected are correct and functioning as intended:
- Button
- Dropdown Button (& Split Button) 
- Hyperlink
- List
- Paginator
- Section Header
- Show/Hide
- Switch
- Table
- Tag 

\
\
Verified by @neilpfeiffer, ready to merge after engineering review.

**Pass** |
-----------|
UX Review|

File: packages/terra-paginator/src/Paginator.module.scss (hunk 1 above)
Comment (emilyrohrbough) [reply packages/terra-paginator/src/Paginator.module.scss]: Can we apply this to any place `/* stylelint-disable max-nesting-depth */` is used? I noticed quite a few still need cleaned up

File: packages/terra-button/CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+Changed
+
```
Comment (supreethmr) [inline packages/terra-button/CHANGELOG.md:6]: Don't we need to add ` Changed hoverstyles to apply only on non-touch devices` (OR something relevant) also in change log to keep track of changes made in css..??

Comment (saket2403) [reply packages/terra-button/CHANGELOG.md:6]: That was done in another PR

**Pull request merged**